### ISSUE TITLE
Train NN and run it through FINN compiler

### ISSUE THREAD

**Issue by kodachi77** [cu-ecen-aeld/final-project-kodachi77#4]
Implement NN architecture using PyTorch, train it on MNIST dataset and run through FINN compiler.
This task includes following:

1. Implement training loop using PyTorch
2. Use Brevis to quantize network
3. Use Tensorboard for monitoring.
4. Adjust hyperparameters as needed to achieve 93-95+% accuracy
5. Create FINN FPGA flow script to compile PTH -> ONNX -> HW
6. Run Verilog testbench to ensure HW is working

**DOD**

- [x] Network achieves at least 93-95% accuracy
- [x] HW bitstream is produced

**Comment by kodachi77** [comment]
Currently network achieved 97+% percent accuracy

See this [commit](https://github.com/kodachi77/zybo-z7-mnist/commit/eab75044fef41d3cae3b11ea5554d3767e717c90)

![nn_training](https://github.com/user-attachments/assets/e310fb43-6dea-49d5-8419-0923f3e6aabf)

**Comment by kodachi77** [comment]
bitstream was generated successfully.

See this [commit](https://github.com/kodachi77/zybo-z7-mnist/commit/aea56424d148a022999bf74000e8e7a1568e6bbc)

![image](https://github.com/user-attachments/assets/f51e1541-1f7d-468a-a5c3-c01ffbba5f1b)

**Comment by kodachi77** [comment]
Simulation completed successfully

![image](https://github.com/user-attachments/assets/68c9fd59-35ec-4a34-9003-ab7558b3df95)

**Comment by kodachi77** [comment]
Issue was in the folding config.

**Comment by kodachi77** [comment]
As an example of the importance of hyperparameter tuning. One of the first runs achieved only ~70% accuracy 

![Image](https://github.com/user-attachments/assets/1339c31b-2285-4cfd-97d5-95fcd1e42cce)
